File: tests/create_new_vault.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
"""
    :copyright: Copyright 2023 by the FreePDM team
    :license:   MIT License.
"""

# Script for creating a new PDM vault.

import os
from os import path
import sys
from pathlib import Path
import appdirs

sys.path.append(os.fspath(Path(__file__).resolve().parents[1]))

from src.filesystem.filesystem import FileSystem
from src.skeleton.config import conf


def clear():
    if os.name == 'nt':
        os.system('cls')
    else:
        os.system('clear')


if __name__ == "__main__":
    clear()
    print("")
    print("Welcome to the create new vault program.")
    print("")
    print("Now a small questionaire is coming for the creation of the new vault.")
    print("For more info look at https://github.com/grd/FreePDM/tests/README.md")
    print("")
    print("The necessary information is the user uid, the vault uid,")
    print("the directory of the mounted sshfs drive and the name of the new vault.")
    print("")
    print("Are you ready? [Y/n]")
    mounting_point = input()
    if mounting_point.lower() == "n":
        exit("Program terminated.")
    
    print("Enter the user name")
    user_name = input()

    print("Enter the user uid")
    user_uid = int(input())

    print("Enter the vault uid")
    vault_uid = int(input())

    print("Enter mounting point of the sshfs directory, e.g. '/mnt/test'")

    while True:
        mounting_point = input()
        if not path.isdir(mounting_point):
            print("Path doesn't exist. Try again.")
        else:
            break

    os.chdir(mounting_point)
    
    if len(os.listdir()) > 0:
        print("")
        print("This is a list of exisisting vaults:")
        for file in os.listdir():
            print(file)

    print("")
    print("Input the directory of your new vault...")
    vault_dir = input()
    
    if path.isdir(vault_dir):
        sys.exit("The vault " + vault_dir + " already exist.")
    
    os.mkdir(vault_dir)
    os.chown(vault_dir, user_uid, vault_uid)

    os.chdir(vault_dir)

    # creating four files: 'All Files.txt', 'FileLocation.txt',
    # 'IndexNumber.txt' and 'Locked.txt'

    all_files = "All Files.txt"
    file_location = "FileLocation.txt"
    index_number = "IndexNumber.txt"
    locked_file = "Locked.txt"

    open(all_files, 'a').close()
    os.chown(all_files, user_uid, vault_uid)

    open(file_location, 'a').close()
    os.chown(file_location, user_uid, vault_uid)

    open(locked_file, 'a').close()
    os.chown(locked_file, user_uid, vault_uid)

    with open(index_number, 'w') as file:
        file.write("0")
    os.chown(index_number, user_uid, vault_uid)

    os.mkdir("PDM")
    os.chown('PDM', user_uid, vault_uid)

    print("Four files have been created: 'All Files.txt', 'FileLocation.txt', ")
    print("'IndexNumber.txt' and 'Locked.txt', and the directory 'PDM")
    print("")
    print(os.listdir())
    print("")
    print("If that is correct, then the vault has been created.\n\n")
    print("Creating a configuration file...\n\n")

    appname = 'FreePDM'
    config_dir = appdirs.user_config_dir(appname)
    config_name = os.path.join(config_dir, 'FreePDM.conf')

    if path.isfile(config_name):
        print("Configuration file already exist.\n\n")
        print("Changing the vault directory...")
        item_list = []
        with open(config_name, "r") as file:
            item_list = file.readlines()

        idx = len(item_list) - 1
        while idx > 0:
            item = item_list[idx]
            if item.startswith("vault"):
                item_list[idx] = "vault = " + "/vault/" + vault_dir + "\n"
            idx -= 1

        with open(config_name, "w") as file:
            for item in item_list:
                file.write(item)
    else:
        os.mkdir(config_dir)
        with open(config_name, "w") as file:
            file.write("[DEFAULT]\n")
            file.write("log_file = \n")
            file.write("logging_is_on = False\n")
            file.write("mounting_point = \"" + mounting_point + "\"\n")
            file.write("server_path = \"" + "/vault/" + vault_dir + "\"\n\n")
            file.write("[user]\n")
            file.write("vault = " + vault_uid + "\n")
            file.write(user_name + " = " + str(user_uid) + "\n")



    print("Program ended successfully.")

File: src/bin/create_vault/create_new_vault.go
// Copyright 2023 The FreePDM team. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package main

import (
	"fmt"
	"log"
	"os"
	"strings"

	ex "github.com/grd/FreePDM/src/extras"
)

// Script for creating a new PDM vault.

func setup() {
	fmt.Println("")
	fmt.Println("Welcome to the create new vault program.")
	fmt.Println("")
	fmt.Println("Now a small questionaire is coming for the creation of the new vault.")
	fmt.Println("For more info look at https://github.com/grd/FreePDM/tests/README.md")
	fmt.Println("")
	fmt.Println("The necessary information is the user uid, the vault uid,")
	fmt.Println("and the name of the new vault.")
	fmt.Println("")
	fmt.Println("Are you ready? [Y/n]")
	var mounting_point string
	fmt.Scan(&mounting_point)
	if strings.ToLower(mounting_point) == "n" {
		fmt.Println("Program terminated.")
		os.Exit(-1)
	}

	fmt.Println("\nEnter the user name")
	var user_name string
	fmt.Scan(&user_name)

	fmt.Println("\nEnter the user uid")
	var user_uid int
	fmt.Scan(&user_uid)

	fmt.Println("Enter the vault uid")
	var vault_uid int
	fmt.Scan(&vault_uid)

	os.Chdir("/vault")

	dir, err := os.ReadDir(".")
	ex.CheckErr(err)
	if len(dir) > 0 {
		fmt.Println("")
		fmt.Println("This is a list of exisisting vaults:")
		for _, file := range dir {
			fmt.Println(file)
		}
	}

	fmt.Println("\nInput the directory of your new vault...")
	var vault_dir string
	fmt.Scan(&vault_dir)

	if ex.DirExists(vault_dir) == true {
		log.Fatalf("The vault \"%v\" already exist.", vault_dir)
	}
	err = os.Mkdir(vault_dir, 0777)
	ex.CheckErr(err)
	err = os.Chown(vault_dir, user_uid, vault_uid)
	ex.CheckErr(err)

	err = os.Chdir(vault_dir)
	ex.CheckErr(err)

	// creating three files: "FileList.txt",
	// "IndexNumber.txt" and "Locked.txt"

	fileListTxt := "FileList.txt"
	index_number := "IndexNumber.txt"
	locked_file := "Locked.txt"

	f1, err := os.Create(fileListTxt)
	f1.Close()
	ex.CheckErr(err)
	os.Chown(fileListTxt, user_uid, vault_uid)

	_, err = os.Create(locked_file)
	ex.CheckErr(err)
	os.Chown(locked_file, user_uid, vault_uid)

	d1 := []byte("0")
	err = os.WriteFile(index_number, d1, 0644)
	ex.CheckErr(err)
	os.Chown(index_number, user_uid, vault_uid)

	err = os.Mkdir("PDM", 0777)
	ex.CheckErr(err)
	err = os.Chown("PDM", user_uid, vault_uid)
	ex.CheckErr(err)

	fmt.Println("Three files have been created: FileList.txt,")
	fmt.Println("IndexNumber.txt and Locked.txt, and the directory PDM")
	fmt.Println("")
	fileList, err := os.ReadDir(".")
	ex.CheckErr(err)
	for _, list := range fileList {
		fmt.Println(list.Name())
	}

	fmt.Println("")
	fmt.Println("If that is correct, then the vault has been created.")

	fmt.Println("Program ended successfully.")
}

func main() {
	setup()
}
